sycl/include/sycl/ext/oneapi/weak_object.hpp

```diff
@@ -0,0 +1,128 @@
+//==-------------- weak_object.hpp --- SYCL weak objects -------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/buffer.hpp>
+#include <sycl/detail/defines_elementary.hpp>
+#include <sycl/ext/oneapi/weak_object_base.hpp>
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext {
+namespace oneapi {
+namespace detail {
+// Helper for creating ranges for empty weak_objects.
+template <int Dims> static range<Dims> createDummyRange() {
+  static_assert(Dims > 0 && Dims < 4, "Invalid dimensionality in range.");
+  if constexpr (Dims == 1)
+    return {1};
+  else if constexpr (Dims == 2)
+    return {1, 1};
+  else
+    return {1, 1, 1};
+}
+} // namespace detail
+
+// The weak_object class.
+// Since we use std::shared_ptr to implementations as a way to handle common
+// reference semantics in SYCL classes, the implementation of the weak_object
+// class holds a weak_ptr to the implementations. The weak_ptr is in the
+// weak_object_base class.
+template <typename SYCLObjT>
+class weak_object : public detail::weak_object_base<SYCLObjT> {
+public:
+  using object_type = typename detail::weak_object_base<SYCLObjT>::object_type;
+
+  constexpr weak_object() noexcept : detail::weak_object_base<SYCLObjT>() {}
+  weak_object(const SYCLObjT &SYCLObj) noexcept
+      : detail::weak_object_base<SYCLObjT>(SYCLObj) {}
+  weak_object(const weak_object &Other) noexcept = default;
+  weak_object(weak_object &&Other) noexcept = default;
+
+  weak_object &operator=(const SYCLObjT &SYCLObj) noexcept {
+    this->MObjWeakPtr = sycl::detail::getSyclObjImpl(SYCLObj);
```

Comment:
Why `getSyclObjImpl` and not `getSyclWeakObjImpl`? If not a bug, I'd like to see a brief comment inside the source code.

Reply:
`getSyclWeakObjImpl` is for getting the impl inside a `weak_object`, not to get the weak variant of impls inside SYCL objects. I can see where it makes sense to add a comment.

Reply:
I have also added a comment about what is happening.